### PR TITLE
[export] Add ir spec

### PR DESCRIPTION
**Pull request by angelayi**
Summary: Copied IR spec over from Executorch

Test Plan: _docs_

Differential Revision: D49829187

### CODE REVIEW

**Review by ydwu4 — APPROVED** [review]

File: docs/source/export.ir_spec.rst [hunk 1]
```diff
@@ -0,0 +1,501 @@
+.. _export.ir_spec:
+
+torch.export IR Specification
+=============================
+
+Export IR is an intermediate representation (IR) for compilers, which bears
+similarities to MLIR and TorchScript. It is specifically designed to express the
+semantics of PyTorch programs that are written in Python. Export IR primarily
```
Comment (gmagogsfm) [inline docs/source/export.ir_spec.rst:8]:
```suggestion
semantics of PyTorch programs. Export IR primarily
```

File: docs/source/export.ir_spec.rst [hunk 2]
```diff
@@ -0,0 +1,501 @@
...
+     target: [str|Callable]
+     args: List[object]
+     kwargs: Dict[str, object]
+     meta: Dict[str, object]
+
+**FX Text Format**
+
+As in the example above, notice that each line has this format::
+
+   %<name>:[...] = <op_name>[target=<target>](args = (%arg1, %arg2, arg3, arg4, …)), kwargs = {})
```
Comment (gmagogsfm) [inline docs/source/export.ir_spec.rst:148]: Is this still accurate? I think a while ago we were talking about supporting kwargs?

File: docs/source/export.ir_spec.rst [hunk 3]
```diff
@@ -0,0 +1,501 @@
...
+^^^^^^^^^^^^^
+
+A ``call_function`` node represents a call to an operator.
+
+**Definitions**
+
+- **Functional:** We say a callable is “functional” if it satisfies all the
+  following requirements:
+
+  - Non-aliasing: The output tensors do not share data storage with each other
```
Comment (gmagogsfm) [inline docs/source/export.ir_spec.rst:193]: Aliasing is allowed now, just not mutations. Let's remove this bullet point.

File: docs/source/export.ir_spec.rst [hunk 4]
```diff
@@ -0,0 +1,501 @@
...
+  :class:`torch.nn.Linear` module call would contain :class:`torch.nn.Linear` in
+  their ``source_fn``, and a node containing the ``addmm`` op from a
+  :class:`torch.nn.functional.Linear` module call would contain
+  :class:`torch.nn.functional.Linear` in their ``source_fn``.
+
+placeholder
+^^^^^^^^^^^
+
+Placeholder represents an input to a graph. Its semantics are exactly the same as in FX.
+Placeholder nodes must be the first N nodes in the nodes list of a graph. N can be zero.
```
Comment (gmagogsfm) [inline docs/source/export.ir_spec.rst:265]: N can be zero?

File: docs/source/export.ir_spec.rst [hunk 5]
```diff
@@ -0,0 +1,501 @@
...
+to be returned.
+
+**Metadata**
+
+Output node has the same metadata as ``call_function`` nodes.
+
+get_attr
+^^^^^^^^
+
+``get_attr`` nodes represent reading a submodule from the encapsulating
```
Comment (gmagogsfm) [inline docs/source/export.ir_spec.rst:306]: And constant tensors for now?

File: docs/source/export.ir_spec.rst [hunk 6]
```diff
@@ -0,0 +1,501 @@
...
+   * - Tuple
+     - Python tuple
+   * - List
+     - Python list
+   * - Dict
+     - Python dict with Scalar keys
+   * - NamedTuple
+     - Python namedtuple
+   * - Custom class
+     - Any custom class defined with ``_register_pytree_node``
```
Comment (gmagogsfm) [inline docs/source/export.ir_spec.rst:471]: dataclass is missing here

File: docs/source/export.ir_spec.rst [hunk 7]
```diff
@@ -0,0 +1,501 @@
...
+
+Memory Formats
```
Comment (gmagogsfm) [inline docs/source/export.ir_spec.rst:473]: Do we need to document this? memory format produced are up to the aten ops semantics I think.

@tugsbayasgalan @bdhirsh does aot_export enforce any particular memory format at all?

cc. @kimishpatel

File: docs/source/export.ir_spec.rst (hunk 5 above)
Comment (angelayi) [reply docs/source/export.ir_spec.rst:306]: No? Constant tensors are lifted

File: docs/source/export.ir_spec.rst (hunk 7 above)
Comment (angelayi) [reply docs/source/export.ir_spec.rst:473]: Oh yeah, we can remove it since dim_order wont be ready by PTC

File: docs/source/export.ir_spec.rst [hunk 8]
```diff
@@ -0,0 +1,471 @@
...
+   * - Tuple
+     - Python tuple
+   * - List
+     - Python list
+   * - Dict
+     - Python dict with Scalar keys
+   * - NamedTuple
+     - Python namedtuple
+   * - Dataclass
+     - Must be registered through ``register_dataclass``
```
Comment (gmagogsfm) [inline docs/source/export.ir_spec.rst:469]: Consider change this (and other func/class/method name) to a reference in rst

**Review by gmagogsfm — APPROVED** [review]

**Pull request closed without merging**